### PR TITLE
Add 3D get_value to Shapes

### PR DESCRIPTION
**Pull request by kevinyamauchi**
# Description
This PR adds the ability to select shapes in 3D by implementing `Shapes._get_value_3d`. To do so, we find all shapes that are intersected by the cursor ray and then return the index of the one closest to the starting point of the cursor ray (i.e., closest to the user). Additionally, we return the coordinates of the intersection, which allows for further interaction (e.g., placing points on planes).

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by kevinyamauchi** [comment]
This is extra WIP, but the main functionality is there. See below for a demo of placing points on the surface of a shape (note this technique would also work for meshes).


https://user-images.githubusercontent.com/1120672/127898013-13c50ce0-edea-4c9a-83c1-e38c5316cc24.mov

The mouse callback:
```python
@shapes_layer.mouse_drag_callbacks.append
def on_click(layer, event):
    shape_index, intersection_point = layer.get_value(
        position=event.position,
        view_direction=event.view_direction,
        dims_displayed=event.dims_displayed,
        world=True
    )

    if (shape_index is not None) and (intersection_point is not None):
        points_layer.add(intersection_point)
```

<details>
  <summary>Example code</summary>
  
  ```python
import napari
import numpy as np

shapes_data = np.array(
    [
        [
            [0, 50, 75, 75],
            [0, 50, 125, 75],
            [0, 100, 125, 125],
            [0, 100, 75, 125]
        ],
        [
            [0, 10, 75, 75],
            [0, 10, 125, 75],
            [0, 40, 125, 125],
            [0, 40, 75, 125]
        ],
        [
            [1, 100, 75, 75],
            [1, 100, 125, 75],
            [1, 50, 125, 125],
            [1, 50, 75, 125]
        ]
    ]
)

properties = {'index': [0, 1, 2]}
viewer = napari.view_points(ndim=3, size=3)
shapes_layer = viewer.add_shapes(
    shapes_data,
    face_color=['magenta', 'green', 'blue'],
    edge_color='white',
    blending='additive',
    properties=properties,
    text='index'
)
points_layer = viewer.layers[0]


@shapes_layer.mouse_drag_callbacks.append
def on_click(layer, event):
    shape_index, intersection_point = layer.get_value(
        position=event.position,
        view_direction=event.view_direction,
        dims_displayed=event.dims_displayed,
        world=True
    )

    if (shape_index is not None) and (intersection_point is not None):
        points_layer.add(intersection_point)


viewer.dims.ndisplay = 3
napari.run()
  ```
</details>

File: napari/layers/shapes/shapes.py
Comment (kevinyamauchi) [inline napari/layers/shapes/shapes.py]: We still need to embed the intersection coordinates in the full nD vector

File: napari/utils/geometry.py
Comment (kevinyamauchi) [inline napari/utils/geometry.py]: I am not sure we will be able to nicely handle doing one plane and an array of planes in the same function. I had to do this funky stuff because the `scale_factor` is a scalar value for the one plane case and is an array of values for the many planes case. Perhaps with some more clever numpy we can get everything to broadcast nicely. Otherwise, we can just make two functions.

File: napari/layers/shapes/_shape_list.py
Comment (kevinyamauchi) [inline napari/layers/shapes/_shape_list.py:792]: Our shape selection operates on the meshes, so I _think_ this should work for all shape types, but I need to test.

**Comment by alisterburt** [comment]
😍 oh my!!!! amazing, love the video! Will pull and have a play later on, so exciting!!!

File: napari/utils/geometry.py
Comment (brisvag) [reply napari/utils/geometry.py]: `np.atleast_2d`?

**Comment by kevinyamauchi** [comment]
I think this is ready for review! A couple of discussion items:

### get_value_and_intersection()
I added a public method `Shapes.get_value_and_intersection()` that returns both the index of the first shape under the cursor and the coordinate of the intersection. My rationale for this is twofold:

1. we have to calculate the intersection points to get the index of the shape since we need to determine which is "on top" (i.e., most in the foreground)
2. I think many people will want both the index and intersection point. For example, the callback to add points on the surface of nD shapes now looks like this (see the `add_points_to_nd_shapes.py` example I added):

```python
@shapes_layer.mouse_drag_callbacks.append
def on_click(layer, event):

    start_point, end_point = layer.get_ray_intersections(
        event.position,
        event.view_direction,
        event.dims_displayed
    )
    if (start_point is not None) and (end_point is not None):
        shape_index, intersection_point = layer.get_value_and_intersection(
            start_point=start_point,
            end_point=end_point,
            dims_displayed=event.dims_displayed
        )

        if (shape_index is not None) and (intersection_point is not None):
            points_layer.add(intersection_point)
```

A couple of question:
- Do we want to add this method? In terms of other layers, I think it would make sense to add it to Labels (the coordinate of the surface of the first label encountered), Image (the coordinate of the max/min/surface encountered), and Mesh (coordinate of the face the mouse intersects)
- Should there be a convenience method that takes the mouse event object? In this case the callback would look like:

```python
@shapes_layer.mouse_drag_callbacks.append
def on_click(layer, event):

    _, intersection_point = layer.get_value_and_intersection_from_event(
        event
    )

    if intersection_point is not None:
        points_layer.add(intersection_point)
```

## interactive editing
I haven't added the interactive editing (e.g., selecting shapes, moving vertices) in this PR. I think there is a lot of value in adding `get_value()` in 3D to all of the layers first so that developers can play with the 3D interactivity. Also, I think we need to first add `get_value()` so that we can experiment with different ways to do the selection, dragging, and adding of elements in 3D. Thoughts?

File: examples/add_points_on_nD_shapes.py
Comment (alisterburt) [inline examples/add_points_on_nD_shapes.py]: is it a value or an index? I wonder if `get_index_and_intersection` is more correct - I see why you would keep with `get_value` in keeping with the rest of the layers though

File: examples/add_points_on_nD_shapes.py
Comment (alisterburt) [inline examples/add_points_on_nD_shapes.py]: as discussed - providing the convenience method which takes the mouse event directly will tidy this up a lot, we should definitely do that!

File: napari/layers/base/base.py
Comment (alisterburt) [inline napari/layers/base/base.py:1039]: good catch, 3D view of 2D data broke this I guess?

File: napari/layers/base/base.py
Comment (alisterburt) [inline napari/layers/base/base.py:1046]: agree with this refactor 👍🏼

File: napari/layers/shapes/_shape_list.py
Comment (alisterburt) [inline napari/layers/shapes/_shape_list.py:792]: I've been calling rays `lines` in the geometry stuff (line-plane intersections etc) - I don't mind line vs ray but we should aim to be consistent

File: napari/layers/shapes/_shape_list.py
Comment (alisterburt) [inline napari/layers/shapes/_shape_list.py:841]: as above, ray vs line

File: napari/layers/shapes/_tests/test_shapes.py
Comment (alisterburt) [inline napari/layers/shapes/_tests/test_shapes.py]:
```suggestion
        assert value == expected
```

File: napari/layers/shapes/shapes.py
Comment (alisterburt) [inline napari/layers/shapes/shapes.py]: I'd prefer `get_index_and_intersection` as discussed

File: napari/utils/geometry.py
Comment (alisterburt) [inline napari/utils/geometry.py:395]: I love this vectorised function in general - I wonder if we shouldn't try to use [einops](https://github.com/arogozhnikov/einops/blob/master/docs/1-einops-basics.ipynb) for some of this stuff, I feel like it makes the code much more concise/maintainable - will open an issue to discuss how people would feel about adding the dependency

File: napari/utils/geometry.py
Comment (alisterburt) [inline napari/utils/geometry.py]: `intersect_ray_with_traingles()`?

File: napari/utils/geometry.py
Comment (alisterburt) [inline napari/utils/geometry.py]: `ray_in_triangles_3d`?

**Review by alisterburt — COMMENTED** [review]
few small comments but this is super close! I fully agree that we should provide this method (with a possible rename suggested) and that getting these `get_value_3d()` methods implemented should be our priority before moving forward on more 3D interactivity.

I have some useful stuff in the pipeline over at #3152 for finding the projection of drag vectors onto arbitrary vectors in displayed data coordinates, I think this will be a good stepping stone for more interactivity when we have the `get_value_3d` implementations in place

This should be able to go in fairly quickly I would say, probably after you've taken a pass over these comments!

File: napari/layers/base/base.py
Comment (kevinyamauchi) [reply napari/layers/base/base.py:1039]: Yes, exactly!

File: napari/utils/geometry.py
Comment (kevinyamauchi) [reply napari/utils/geometry.py:395]: Einops definitely looks interesting. for some things, [np.einsum](https://numpy.org/doc/stable/reference/generated/numpy.einsum.html) might suffice. Although for this stuff, my guess is that the vectorized version is faster, which for mouse callbacks that get called on cursor move is a key consideration.

File: napari/layers/shapes/_tests/test_shapes.py
Comment (kevinyamauchi) [reply napari/layers/shapes/_tests/test_shapes.py]: oops, good catch!

File: napari/layers/shapes/_shape_list.py
Comment (kevinyamauchi) [reply napari/layers/shapes/_shape_list.py:792]: Good point. To me, rays and lines are different: rays have a starting point and extend in a single direction, whereas lines extend in both directions from a point. In that frame, we should probably rename a lot of these things from ray to line. Thoughts?

File: napari/layers/shapes/shapes.py
Comment (kevinyamauchi) [reply napari/layers/shapes/shapes.py]: Agreed. Done!

File: napari/layers/shapes/_shape_list.py
Comment (kevinyamauchi) [reply napari/layers/shapes/_shape_list.py:792]: Where appropriate, I have renamed everything in the geometry utils that is currently "ray" to "line".

**Comment by kevinyamauchi** [comment]
Thanks for the review, @alisterburt ! I believe I have addressed your comments.

**Review by alisterburt — APPROVED** [review]
fantatic, great work @kevinyamauchi ! Let's get this in

File: examples/add_points_on_nD_shapes.py
Comment (sofroniewn) [inline examples/add_points_on_nD_shapes.py:1]: This is so much fun!!!

File: napari/layers/shapes/_shape_list.py
Comment (sofroniewn) [reply napari/layers/shapes/_shape_list.py:792]: agreed on the ray vs line definition from @kevinyamauchi above - not sure what that means for our current naming though - consistency is probably the most important thing though

**Review by sofroniewn — APPROVED** [review]
LGTM too! @alisterburt merge away when ready!!

File: napari/layers/shapes/_shape_list.py
Comment (kevinyamauchi) [reply napari/layers/shapes/_shape_list.py:792]: I have renamed everything in the geometry utils to match the definition I gave above, so I believe we are now self-consistent.

**Comment by kevinyamauchi** [comment]
Thanks for the review @sofroniewn ! I'd like to just update the `get_index_and_intersection_from_event()` method to take the properties from the mouse event as kwargs instead of the event object (as we discussed in the dev meeting yesterday). After that, I think we are good to go!

File: napari/layers/shapes/_shape_list.py
Comment (sofroniewn) [reply napari/layers/shapes/_shape_list.py:792]: Nice, will merge now then!!

**Comment by sofroniewn** [comment]
> I'd like to just update the get_index_and_intersection_from_event() method to take the properties from the mouse event as kwargs instead of the event object (as we discussed in the dev meeting yesterday). After that, I think we are good to go!

ah ok, nice - well recalled!

File: examples/add_points_on_nD_shapes.py
Comment (kevinyamauchi) [reply examples/add_points_on_nD_shapes.py]: After the discussion in the euro dev meeting yesterday, I will refactor the convenience method to take the properties of the mouse event as kwargs instead of the event object itself.

File: examples/add_points_on_nD_shapes.py
Comment (kevinyamauchi) [reply examples/add_points_on_nD_shapes.py]: We chatted about this on zulip, but posting an answer here for the record. I have added a method called `get_index_and_intersection()` which returns both the index and intersection of the first shape encountered along the click ray. We are keeping `get_value` as is to be consistent across the layers.

**Comment by kevinyamauchi** [comment]
I have refactored `shapes.get_index_and_intersection()` to take the properties of the mouse event rather than the mouse event.

@alisterburt , would you like to taking another look and then merging if you think it's ready?

edit: here is how the mousecallback looks now

```python
@shapes_layer.mouse_drag_callbacks.append
def on_click(layer, event):

    shape_index, intersection_point = layer.get_index_and_intersection(
        event.position,
        event.view_direction,
        event.dims_displayed
    )

    if (shape_index is not None) and (intersection_point is not None):
        points_layer.add(intersection_point)
```

File: examples/add_points_on_nD_shapes.py
Comment (alisterburt) [inline examples/add_points_on_nD_shapes.py:55]: I understand the motivation for this refactor but really do think this API is messier, it forces the user to be aware of
- the structure of the event object
- the dims model

I think if we want to keep this knowledge of the events off the layer (which I understand and support) then a better approach is to provide all interactivity utilities in an `interactivity_utils.py` file, have those easily accessible and always point to those

will not hold up merge for this, we can get this and click and drag in then refactor when all of the pieces are together

File: napari/utils/_tests/test_geometry.py
Comment (alisterburt) [inline napari/utils/_tests/test_geometry.py:86]: we should refactor so that the only func we have for this is a vectorised line-multiple plane intersection which returns the squeezed intersection point if only 1 is returned - again, not blocking for this, just noting things 🙂

**Review by alisterburt — APPROVED** [review]
Approving and merging! Left a few comments for things we can think about moving forward but let's get the functionality in 🎉

File: examples/add_points_on_nD_shapes.py
Comment (sofroniewn) [reply examples/add_points_on_nD_shapes.py:55]: ok yeah, we can continue to iterate here in future PRs

**Pull request merged**
